### PR TITLE
Send a widget ID with Shift-F6 for automated testing

### PR DESCRIPTION
**Pull request by shundhammer**
https://hackweek.suse.com/16/projects/make-yast-testing-independent-of-keyboard-shortcuts

and

https://fate.suse.com/324098

NCurses part

### CODE REVIEW

File: VERSION.cmake [hunk 1]
```diff
@@ -1,6 +1,6 @@
-SET( VERSION_MAJOR "2" )
-SET( VERSION_MINOR "48" )
-SET( VERSION_PATCH "4" )
+SET( VERSION_MAJOR "4" )
+SET( VERSION_MINOR "0" )
+SET( VERSION_PATCH "0" )
```
Comment (jreidinger) [inline VERSION.cmake]: @mvidner is somewhere versioning documented? I am really not sure about this.

File: src/NCDialog.cc [hunk 2]
```diff
@@ -1095,6 +1096,24 @@ void NCDialog::processInput( int timeout_millisec )
...
 		break;
 
+	    case KEY_F( 18 ): // = Shift-F6 on e.g. a linux console
+                {
+                    yuiMilestone() << "Asking for widget ID" << std::endl;
+                    YWidget * widget = YNCursesUI::ui()->askSendWidgetID();
+
+                    if ( widget )
+                    {
+                        NCPushButton * button = dynamic_cast<NCPushButton *>( widget );
```
Comment (jreidinger) [inline src/NCDialog.cc]: so only buttons can be selected? not radio buttons and others? I think we should at least log if widget with given ID exists, just it is not button, so we do nothing.

File: src/NCPopupTextEntry.h [hunk 3]
```diff
@@ -58,6 +58,12 @@ class NCPopupTextEntry : public NCPopup
     void   setValue( const std::string & ntext ) { wtext->setValue( ntext ); }
 
     std::string value() { return wtext->value(); }
+
+    static std::string askForText( const wpos at,
+                                   const std::string & label,
+                                   const std::string & text,
```
Comment (jreidinger) [inline src/NCPopupTextEntry.h]: maybe make text default param `""`? It looks for me as usual case, where you ask for something.

File: src/NCDialog.cc (hunk 2 above)
Comment (shundhammer) [reply src/NCDialog.cc]: The first thing that happens is that the widget gets the keyboard focus. This is done for all widgets. But if it's a button, it is also immediately activated just as if the user pressed [Return].

File: src/NCPopupTextEntry.h (hunk 3 above)
Comment (shundhammer) [reply src/NCPopupTextEntry.h]: Yes, we can do that.

**Comment by shundhammer** [comment]
After discussing this with @lslezak, now using less aggressive versioning: Bumped the minor version of both package and library.

The Travis build failure is expected for the time being since this UI now requires the latest libyui with that new `sendWidgetID()` method.

File: src/YNCursesUI.cc [hunk 4]
```diff
@@ -423,3 +424,24 @@ void YNCursesUI::setConsoleFont( const std::string & console_magic,
     }
 }
 
+
+YWidget * YNCursesUI::askSendWidgetID()
+{
+    std::string id = NCPopupTextEntry::askForText( wpos( 0, 0 ),
+                                                   "Widget ID:", // label
```
Comment (lslezak) [inline src/YNCursesUI.cc]: NP: The label should be consistent with Qt which uses "Enter Widget ID:" label (https://github.com/libyui/libyui-qt/pull/74/files/bb0fa0426b545c8740007bba8c14a7d1cfbd91d9#diff-0ffb2fa84e56876f016657d53238482eR625)

Plus should be marked for translation.

File: src/YNCursesUI.cc [hunk 5]
```diff
@@ -423,3 +424,24 @@ void YNCursesUI::setConsoleFont( const std::string & console_magic,
...
+
+    if ( ! id.empty() )
+    {
+        try
+        {
+            return sendWidgetID( id );
+        }
+        catch ( YUIWidgetNotFoundException & ex )
+        {
+            YUI_CAUGHT( ex );
```
Comment (lslezak) [inline src/YNCursesUI.cc]: Um, it should display an error popup, like the Qt UI: https://github.com/libyui/libyui-qt/pull/74/files/bb0fa0426b545c8740007bba8c14a7d1cfbd91d9#diff-0ffb2fa84e56876f016657d53238482eR643

Comment (shundhammer) [reply src/YNCursesUI.cc]: I was considering that, but we don't seem to have a simple pre-made function for that, unlike in Qt.

Comment (lslezak) [reply src/YNCursesUI.cc]: OK, displaying a popup is not trivial in NCurses UI, so let's skip it here.

**Pull request merged**